File: clients/ios/Lextures/Core/Auth/AuthSession.swift
import Foundation
import Observation

enum AuthPhase: Equatable {
    case splash
    case unauthenticated
    case authenticated
}

enum MFARequired: Equatable {
    case challenge
    case setup
}

@MainActor
@Observable
final class AuthSession {
    var phase: AuthPhase = .splash
    var accessToken: String?
    var userEmail: String?
    var mfaRequired: MFARequired?

    private var refreshToken: String?
    private var refreshTask: Task<Void, Never>?

    var isSignedIn: Bool {
        accessToken != nil
    }

    init() {
        accessToken = KeychainStore.read(key: KeychainStore.Keys.accessToken)
        refreshToken = KeychainStore.read(key: KeychainStore.Keys.refreshToken)
    }

    func finishSplash() {
        phase = isSignedIn ? .authenticated : .unauthenticated
    }

    // MARK: Token refresh
    //
    // Backend access tokens last 15 minutes by design; refresh tokens last 30 days
    // and rotate on every exchange. Refreshing keeps the mobile session alive for
    // weeks without weakening the backend's short-lived bearer tokens.

    /// Refreshes the access token when it is missing, expired, or expiring soon.
    /// Coalesces concurrent callers into one in-flight exchange.
    func refreshIfNeeded(force: Bool = false) async {
        if let task = refreshTask {
            await task.value
            return
        }
        guard refreshToken != nil else { return }
        if !force, let token = accessToken, let expiry = Self.jwtExpiry(token),
           expiry.timeIntervalSinceNow > 120 {
            return
        }

        let task = Task { await performRefresh() }
        refreshTask = task
        await task.value
        refreshTask = nil
    }

    private func performRefresh() async {
        guard let refreshToken else { return }
        do {
            let response = try await AuthAPI.refresh(refreshToken: refreshToken)
            guard let token = response.accessToken, !token.isEmpty else { return }
            try? KeychainStore.save(key: KeychainStore.Keys.accessToken, value: token)
            if let rotated = response.refreshToken, !rotated.isEmpty {
                try? KeychainStore.save(key: KeychainStore.Keys.refreshToken, value: rotated)
                self.refreshToken = rotated
            }
            accessToken = token
            if let email = response.user?.email {
                userEmail = email
            }
        } catch {
            // Only an explicit rejection ends the session; network blips keep it.
            if case APIError.httpStatus(let code, _) = error, code == 401 || code == 403 {
                signOut()
            }
        }
    }

    /// Decodes the `exp` claim from a JWT without verifying the signature.
    static func jwtExpiry(_ jwt: String) -> Date? {
        let segments = jwt.split(separator: ".")
        guard segments.count >= 2 else { return nil }
        var base64 = String(segments[1])
            .replacingOccurrences(of: "-", with: "+")
            .replacingOccurrences(of: "_", with: "/")
        while base64.count % 4 != 0 { base64.append("=") }
        guard
            let data = Data(base64Encoded: base64),
            let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
            let exp = json["exp"] as? TimeInterval
        else { return nil }
        return Date(timeIntervalSince1970: exp)
    }

    func applyTokenResponse(_ response: AuthTokenResponse) throws {
        if response.requiresMFA == true, response.mfaPendingToken != nil {
            mfaRequired = response.mfaSetupRequired == true ? .setup : .challenge
            throw AuthSessionError.mfaRequired
        }

        guard let token = response.accessToken, !token.isEmpty else {
            throw AuthSessionError.missingAccessToken
        }

        try KeychainStore.save(key: KeychainStore.Keys.accessToken, value: token)
        if let refresh = response.refreshToken {
            try? KeychainStore.save(key: KeychainStore.Keys.refreshToken, value: refresh)
            refreshToken = refresh
        }

        accessToken = token
        userEmail = response.user?.email
        mfaRequired = nil
        phase = .authenticated
    }

    func signOut() {
        KeychainStore.deleteAll()
        accessToken = nil
        refreshToken = nil
        userEmail = nil
        mfaRequired = nil
        phase = .unauthenticated
    }

    func serverUnreachableMessage() -> String {
        "Could not reach the server. Is the API running?"
    }

    enum AuthSessionError: LocalizedError {
        case mfaRequired
        case missingAccessToken

        var errorDescription: String? {
            switch self {
            case .mfaRequired:
                return "Multi-factor authentication is required. Complete sign-in on the web app for now."
            case .missingAccessToken:
                return "Unexpected sign-in response."
            }
        }
    }
}

File: clients/android/app/src/main/kotlin/com/lextures/android/core/auth/AuthSession.kt
package com.lextures.android.core.auth

import android.app.Application
import android.util.Base64
import androidx.lifecycle.AndroidViewModel
import androidx.lifecycle.viewModelScope
import com.lextures.android.core.config.AppConfiguration
import com.lextures.android.core.network.ApiError
import com.lextures.android.core.offline.OfflineService
import kotlinx.coroutines.delay
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.launch
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock

enum class AuthPhase {
    Splash,
    Unauthenticated,
    Authenticated,
}

class AuthSession(application: Application) : AndroidViewModel(application) {
    private val tokenStore = TokenStore(application)

    private val _phase = MutableStateFlow(AuthPhase.Splash)
    val phase: StateFlow<AuthPhase> = _phase.asStateFlow()

    private val _accessToken = MutableStateFlow(tokenStore.readAccessToken())
    val accessToken: StateFlow<String?> = _accessToken.asStateFlow()

    private val _userEmail = MutableStateFlow<String?>(null)
    val userEmail: StateFlow<String?> = _userEmail.asStateFlow()

    val isSignedIn: Boolean
        get() = !_accessToken.value.isNullOrBlank()

    // Declared before init: the splash coroutine below can call refreshIfNeeded()
    // synchronously during construction (Main.immediate), and Kotlin runs property
    // initializers and init blocks strictly in declaration order.
    private val refreshMutex = Mutex()

    init {
        viewModelScope.launch {
            // Refresh in parallel with the splash so a stale 15-minute access
            // token is replaced before the app shows.
            val splash = launch { delay(SPLASH_DURATION_MS) }
            refreshIfNeeded()
            splash.join()
            finishSplash()
        }
        viewModelScope.launch {
            // Keep the access token fresh while the app process is alive.
            while (true) {
                delay(REFRESH_INTERVAL_MS)
                if (_phase.value == AuthPhase.Authenticated) {
                    refreshIfNeeded()
                }
            }
        }
    }

    fun finishSplash() {
        _phase.value = if (isSignedIn) AuthPhase.Authenticated else AuthPhase.Unauthenticated
    }

    /** Hook for ON_RESUME: the token has likely expired while backgrounded. */
    fun onAppResumed() {
        if (_phase.value == AuthPhase.Authenticated) {
            viewModelScope.launch { refreshIfNeeded() }
        }
    }

    // Token refresh
    //
    // Backend access tokens last 15 minutes by design; refresh tokens last 30 days
    // and rotate on every exchange. Refreshing keeps the mobile session alive for
    // weeks without weakening the backend's short-lived bearer tokens.

    /** Refreshes the access token when it is missing, expired, or expiring soon. */
    suspend fun refreshIfNeeded(force: Boolean = false) {
        refreshMutex.withLock {
            val refreshToken = tokenStore.readRefreshToken() ?: return
            if (!force) {
                val expiry = jwtExpiryEpochSeconds(_accessToken.value)
                if (expiry != null && expiry - System.currentTimeMillis() / 1000 > 120) return
            }
            try {
                val response = AuthApi.refresh(refreshToken)
                val token = response.accessToken?.takeIf { it.isNotBlank() } ?: return
                tokenStore.saveTokens(token, response.refreshToken ?: refreshToken)
                _accessToken.value = token
                response.user?.email?.let { _userEmail.value = it }
            } catch (e: ApiError.HttpStatus) {
                // Only an explicit rejection ends the session; network blips keep it.
                if (e.code == 401 || e.code == 403) {
                    signOut()
                }
            } catch (_: Exception) {
                // Transport/decoding errors: keep the session and retry later.
            }
        }
    }

    suspend fun applyTokenResponse(response: AuthTokenResponse) {
        if (response.requiresMfa == true && response.mfaPendingToken != null) {
            throw AuthSessionError.MfaRequired
        }

        val token = response.accessToken?.takeIf { it.isNotBlank() }
            ?: throw AuthSessionError.MissingAccessToken

        tokenStore.saveTokens(token, response.refreshToken)
        _accessToken.value = token
        _userEmail.value = response.user?.email
        _phase.value = AuthPhase.Authenticated
    }

    fun signOut() {
        OfflineService.get(getApplication()).clearAllOnLogout()
        tokenStore.clearAll()
        _accessToken.value = null
        _userEmail.value = null
        _phase.value = AuthPhase.Unauthenticated
    }

    fun serverUnreachableMessage(): String =
        "Could not reach the server at ${AppConfiguration.apiBaseUrl}. Is the API running?"

    fun mapError(error: Throwable): String {
        return when (error) {
            is AuthSessionError -> error.message ?: error.toString()
            is ApiError.Transport -> serverUnreachableMessage()
            is ApiError -> error.message ?: error.toString()
            else -> error.localizedMessage ?: error.toString()
        }
    }

    sealed class AuthSessionError(message: String) : Exception(message) {
        data object MfaRequired : AuthSessionError(
            "Multi-factor authentication is required. Complete sign-in on the web app for now.",
        )

        data object MissingAccessToken : AuthSessionError(
            "Unexpected sign-in response.",
        )
    }

    companion object {
        private const val SPLASH_DURATION_MS = 900L
        private const val REFRESH_INTERVAL_MS = 10L * 60 * 1000

        /** Decodes the `exp` claim from a JWT without verifying the signature. */
        fun jwtExpiryEpochSeconds(jwt: String?): Long? {
            if (jwt.isNullOrBlank()) return null
            val segments = jwt.split(".")
            if (segments.size < 2) return null
            return runCatching {
                val payload = Base64.decode(
                    segments[1],
                    Base64.URL_SAFE or Base64.NO_PADDING or Base64.NO_WRAP,
                ).decodeToString()
                Regex(""""exp"\s*:\s*(\d+)""").find(payload)?.groupValues?.get(1)?.toLong()
            }.getOrNull()
        }
    }
}
